### PR TITLE
test PR for ndeg-twisted-clover branch to keep up to date on changes

### PR DESCRIPTION
**Pull request by kostrzewa**
Just a test PR to our fork of the generic_kernel branch to quickly compare what has changed and to be notified of changes as they are made (if desired).

## TODOs 

**a checkmark means that the non-deg version is wired up completely, has a `host_reference` implementation (if applicable) and the test passes**

- [x] `DiracTwistedClover::M` #6 
- [x] `DiracTwistedClover::TwistClover`
- [x] `DiracTwistedClover::DslashXpay`
- [x] `DiracTwistedClover::MdagM`
- [x] `DiracTwistedCloverPC::TwistCloverInv`
- [x] `DiracTwistedCloverPC::Dslash`
- [x] `DiracTwistedCloverPC::DslashXpay`
- [x] `DiracTwistedCloverPC::M`
- [x] `DiracTwistedCloverPC::MdagM`
- [x] `DiracTwistedCloverPC::prepare`
- [x] `DiracTwistedCloverPC::reconstruct`
- [x] `DiracTwistedCloverPC::prefetch`
- [x] `CloverInvert::apply` (adapt to nondeg, see https://github.com/qcdcode/quda/pull/4/commits/7165a1841c6cf56dfd59c62c4b1c331f74e7d3cc)

~~For the PC operator, there's also a bunch of stuff related to the packing kernels which is non-trivial AFAIK.~~ (does not apply to the clover operators)

## stuff for later

- [ ] Go through all operators and correct flops / bytes

Part of `DiracTwistedClover` and `DiracTwistedCloverPC` are the functionalities related to interfacing for the coarse operator. I would propose that in a first round we omit these. However, we have to make sure that the logic disallows a non-deg coarse op to be created...

- [ ] check that `DiracTwistedClover::createCoarseOp` cannot be called for the non-deg operator
- [ ] check that `DiracTwistedCloverPC::createCoarseOp` cannot be called for the non-deg operator

### CODE REVIEW

File: include/kernels/dslash_ndeg_twisted_clover_preconditioned.cuh [hunk 1]
```diff
@@ -7,32 +7,32 @@
 
 namespace quda
 {
-
-  template <typename Float, int nColor, QudaReconstructType reconstruct_, bool dynamic_clover_>
```
Comment (kostrzewa) [inline include/kernels/dslash_ndeg_twisted_clover_preconditioned.cuh]: this template signature now looks different, I think (see `TwistedCloverArg`)

**Comment by urbach** [comment]
> @kostrzewa commented on this pull request.
> 
> 
> 
> > @@ -7,32 +7,32 @@
>  
>  namespace quda
>  {
> -
> -  template <typename Float, int nColor, QudaReconstructType reconstruct_, bool dynamic_clover_>
> 
> this template signature now looks different, I think (see `TwistedCloverArg`)

yes, I already started to change it...

**Comment by kostrzewa** [comment]
Just to keep you up to date, I'm making progress on the test for the nonPC operator and have reached the stage where I'm adjusting the signature of `tmc_ndeg_mat` to take into account that the number of flavours as well as the checkerboarding are properties of the `ColorSpinorField` (such that also the two-flavour operator has only a single input and a single output field, such that the starting indices of the parities and flavours have to be determined at the point of entry).

**Comment by kostrzewa** [comment]
See https://github.com/qcdcode/helper_scripts/blob/master/quda/run_test_NdegTwistedClover.sh for run script.

**Comment by kostrzewa** [comment]
Running `dslash_test` for `NdegTwistedClover`, we get:

```
Calculating reference implementation...done.
Tuning...

CUDA Exception: Warp Illegal Address
The exception was triggered at PC 0x5555757d9550 (gauge_field_order.h:1636)

Thread 1 "dslash_test" received signal CUDA_EXCEPTION_14, Warp Illegal Address.
[Switching focus to CUDA kernel 0, grid 12184, block (20067,0,0), thread (0,0,0), device 0, sm 0, warp 21, lane 0]
quda::dslashGPU<quda::nDegTwistedClover, quda::packShmem, 2, false, true, (quda::KernelType)5, quda::NdegTwistedCloverArg<float, 3, 4, (QudaReconstructType_s)18> ><<<(20736,2,2),(16,1,1)>>> () at /home/bartek/code/quda.ndeg-twisted-clover/lib/../include/color_spinor.h:1225
1225		  y.data[s*Nc + i].y += A(i,j).real() * x.data[s*Nc + j].imag();
```

So I think now would perhaps be a good time to PR this to the QUDA devs who might immediately be able to tell what's going wrong...

The template arguments with which the kernel is instantiated can be identified via `qudaLaunchKernel(dslashGPU<D, P, nParity, dagger, xpay, kernel_type, Arg>, tp, stream, arg);`.

**Comment by urbach** [comment]
Yes, a PR is a good idea! Will you or shall I?
-- 
CU Mobil, www.carsten-urbach.eu

**Comment by kostrzewa** [comment]
I'll open one in the next few days. Fixing this was actually very instructive. You had the right idea commenting on the `in.VolumeCB()` at the point of the volume loop launch in the functor. Turns out `VolumeCB()` is too large by a factor of two on account of our abusing the fifth dimension for flavour. This is why NdegTwistedMass uses `in.getDslashConstant().volume_4d_cb` instead and we have to do the same here.

**Comment by urbach** [comment]
Okay, I see. I had put this in comment because I wanted to investigate this at some point. Now you did already! Very good.

Sorry, I'll probably only be able to look at the things you did on Monday.
-- 
CU Mobil, www.carsten-urbach.eu

**Comment by kostrzewa** [comment]
> Sorry, I'll probably only be able to look at the things you did on Monday.

Sure, I just happened to have the motivation and time yesterday and today to go through all of this.

The test is reasonably close to passing now

```
7134.751081us per kernel call
1337720832 flops per kernel call, 2016 flops per site
GFLOPS = 187.493693
Effective halo bi-directional bandwidth (GB/s) GPU = 0.744023 ( CPU = 0.744031, min = 0.652300 , max = 0.767334 ) for aggregate message size 5308416 bytes
Results: CPU = 7434030.804202, CUDA=7499751.827586, CPU-CUDA = 7499751.827588
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from dslash
[ RUN      ] dslash.verify
0 fails = 108293
1 fails = 107960
2 fails = 108306
3 fails = 107985
4 fails = 108229
5 fails = 107980
6 fails = 108280
7 fails = 107959
8 fails = 108286
9 fails = 108015
10 fails = 108350
11 fails = 107935
12 fails = 108293
13 fails = 107960
14 fails = 108306
15 fails = 107985
16 fails = 108229
17 fails = 107980
18 fails = 108280
19 fails = 107959
20 fails = 108287
21 fails = 108015
22 fails = 108350
23 fails = 107935
1.000000e-01 Failures: 150792 / 31850496  = 4.734369e-03
1.000000e-02 Failures: 2074728 / 31850496  = 6.513958e-02
1.000000e-03 Failures: 2595157 / 31850496  = 8.147933e-02
1.000000e-04 Failures: 2648197 / 31850496  = 8.314461e-02
1.000000e-05 Failures: 2653559 / 31850496  = 8.331296e-02
1.000000e-06 Failures: 2654137 / 31850496  = 8.333110e-02
1.000000e-07 Failures: 2829939 / 31850496  = 8.885070e-02
1.000000e-08 Failures: 22727814 / 31850496  = 7.135780e-01
1.000000e-09 Failures: 30902545 / 31850496  = 9.702375e-01
1.000000e-10 Failures: 31755450 / 31850496  = 9.970159e-01
1.000000e-11 Failures: 31840955 / 31850496  = 9.997004e-01
1.000000e-12 Failures: 31849553 / 31850496  = 9.999704e-01
1.000000e-13 Failures: 31850409 / 31850496  = 9.999973e-01
1.000000e-14 Failures: 31850483 / 31850496  = 9.999996e-01
1.000000e-15 Failures: 31850496 / 31850496  = 1.000000e+00
1.000000e-16 Failures: 31850496 / 31850496  = 1.000000e+00
/home/bartek/code/quda.ndeg-twisted-clover/tests/dslash_test.cpp:926: Failure
Expected: (deviation) <= (tol), actual: 1 vs 0.0001
CPU and CUDA implementations do not agree
[  FAILED  ] dslash.verify (3964 ms)
```

**Comment by urbach** [comment]
I'll review the host imlementation as soo as I can.

**Comment by kostrzewa** [comment]
> I'll review the host imlementation as soo as I can.

Sorry, the "request for review" was really just to get us both subscribed to this issue, please don't feel under time pressure because I happened to work on this yesterday.

**Comment by urbach** [comment]
> > I'll review the host imlementation as soo as I can.
> 
> Sorry, the "request for review" was really just to get us both subscribed to this issue, please don't feel under time pressure because I happened to work on this yesterday.

all fine!

File: include/kernels/dslash_ndeg_twisted_clover.cuh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        
+        Vector tmp;
+#pragma unroll
+        for (int chirality = 0; chirality < 2; chirality++) {
+          constexpr int n = Arg::nColor * Arg::nSpin / 2;
+          HMatrix<real, n> A = arg.A(clover_coord.x_cb, parity, chirality);
+          HalfVector x_chi = x.chiral_project(chirality);
+          HalfVector Ax_chi = A * x_chi;
+          // i * mu * gamma_5 * tau_3
+          const complex<real> b(0.0, (chirality^flavor) == 0 ? static_cast<real>(arg.b) : -static_cast<real>(arg.b));
```
Comment (urbach) [inline include/kernels/dslash_ndeg_twisted_clover.cuh:87]: @kostrzewa do you agree on this line? Before the sign change due to flavor was missing.

Comment (kostrzewa) [reply include/kernels/dslash_ndeg_twisted_clover.cuh:87]: yes, 0^0 = 0, 1^0 = 0^1 = 1, 1^1 = 0

Comment (kostrzewa) [reply include/kernels/dslash_ndeg_twisted_clover.cuh:87]: and as far as I can tell, the sign of mu is not adjusted further up the call graph

File: include/kernels/dslash_ndeg_twisted_clover_preconditioned.cuh (hunk 1 above)
Comment (urbach) [reply include/kernels/dslash_ndeg_twisted_clover_preconditioned.cuh]: Should be fixed now

**Comment by urbach** [comment]
I have to take the decision whether or not to keep the `asymmetric` and `!xpay` versions of the preconditioned kernel. I think I'll first go without now. Any additional thoughts?

**Comment by urbach** [comment]
on the other hand, we would break the rule I suppose...

File: include/kernels/dslash_ndeg_twisted_clover.cuh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+          Ax_chi += b * x_chi;
+          tmp += Ax_chi.chiral_reconstruct(chirality);
+        }
+
+        tmp.toNonRel();
+        // c * tau_1
+        tmp += (flavor == 0 ? arg.c * x1 : arg.c * x0);
```
Comment (kostrzewa) [inline include/kernels/dslash_ndeg_twisted_clover.cuh:94]: @urbach Unlike `b`, The normalisation of `c` should be 1.0, right? It is added to `tmp` as a full `Vector` rather than separately to each chirality.

Wanted to make sure that we don't have a misunderstanding as your write in the Wiki that also `c` shoud have the factor 0.5 to compensate for the reconstruction:

https://github.com/qcdcode/quda/wiki/ndeg-twisted-clover-kernel-and-testing/_compare/c8cb509073ebe5aad563ff8768bd851f996fb884...33d7f42d8bd067a0b805f7a360c8df5de4b75044#diff-b014db1857a953f6c396fc04597381a347d941d3463fe4e3d318d840eb0ee482R9

Comment (kostrzewa) [reply include/kernels/dslash_ndeg_twisted_clover.cuh:94]: Similarly, `a` should also not be scaled.

**Comment by kostrzewa** [comment]
> I have to take the decision whether or not to keep the asymmetric and !xpay versions of the preconditioned kernel. I think I'll first go without now. Any additional thoughts?

I think we will need the `!xpay` version in the HMC, right? Although of course one could also rely on the `xpay = true` version and accumulate a zero zpinor.

image

Of course, much like in the case of `DiracTwistedClover`, we can build this out of a special function applying the inverse twisted clover term and call the Wilson `Dslash` (see `lib/dirac_twisted_clover.cpp` -> `DiracTwistedCloverPC::Dslash`)

File: include/kernels/dslash_ndeg_twisted_clover.cuh (hunk 3 above)
Comment (urbach) [reply include/kernels/dslash_ndeg_twisted_clover.cuh:94]: I agree with you. Didn't I say it is only a facto 1/2 in the chiral basis?

Comment (urbach) [reply include/kernels/dslash_ndeg_twisted_clover.cuh:94]: I have clarified this in the docu.

**Comment by kostrzewa** [comment]
I've put a bunch of TODOs at the top which might help in scoping out the tasks ahead. Not sure if this is the best way of organising things as this PR will get rather full rather quickly I presume.

File: tests/dslash_test.cpp [hunk 4]
```diff
@@ -629,31 +629,15 @@ void dslashRef()
 	tm_matpc(spinorRef->V(), hostGauge, spinor->V(), inv_param.kappa, inv_param.mu, inv_param.twist_flavor, inv_param.matpc_type, dagger, inv_param.cpu_prec, gauge_param);
       else
       {
-        int tm_offset = 12*spinorRef->Volume();
```
Comment (kostrzewa) [inline tests/dslash_test.cpp]: a note one why I changed this (beyond making things cleaner in my opinion): the `tm_offset` is misleading as it refers to `Volume()`, which suggests that the flavours are separated by an entire space-time-volume. However, the factor `12` is wrong here, it should be `24` (complex) and half the volume instead.

**Comment by kostrzewa** [comment]
@urbach I'm currently merging in the latest changes from `feature/generic_kernel` and there is a minor change relevant to the shared memory allocation in the preconditioned kernel which will have to be propagated to the ndeg wisted clover preconditioned kernel.

```
diff --git a/include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh b/include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh
index 0891fa2c1..16a255831 100644
--- a/include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh
+++ b/include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh
@@ -94,7 +94,7 @@ namespace quda
 
       if (isComplete<kernel_type>(arg, coord) && active) {
         if (!dagger || Arg::asymmetric) { // apply A^{-1} to D*in
-          SharedMemoryCache<Vector> cache(device::block_dim());
+          SharedMemoryCache<Vector> cache(target::block_dim());
           // to apply the preconditioner we need to put "out" in shared memory so the other flavor can access it
           cache.save(out);
           cache.sync(); // safe to sync in here since other threads will exit
```

Note also that Kepler support has been dropped.

**Comment by kostrzewa** [comment]
As discussed, here's the PR for build testing: https://github.com/lattice/quda/pull/1121

File: lib/dslash_ndeg_twisted_clover_preconditioned.cu [hunk 5]
```diff
@@ -0,0 +1,156 @@
+#ifndef USE_LEGACY_DSLASH
```
Comment (kostrzewa) [inline lib/dslash_ndeg_twisted_clover_preconditioned.cu]: this can be removed as the legacy dslash has been evicted from the code completely

**Comment by urbach** [comment]
Yes, did so already...
-- 
CU Mobil, www.carsten-urbach.eu

**Comment by kostrzewa** [comment]
@urbach I've added `tmc_ndeg_dslash` which can be used to test the non-degerate `DiracTwistedCloverPC::Dslash`. Maybe we can have a chat one of these days to discuss again.

**Comment by kostrzewa** [comment]
If you want, I can start wiring up the operators `DiracTwistedCloverPC::Dslash`, `DiracTwistedCloverPC::DslashXpay`, `DiracTwistedCloverPC::M` and `DiracTwistedCloverPC::MdagM` to call the appropriate `Apply*` functions (and combinations of other functions) based on the flavour and preconditioning type. I just don't want to cause conflicts with your local copy.

**Comment by urbach** [comment]
> If you want, I can start wiring up the operators `DiracTwistedCloverPC::Dslash`, `DiracTwistedCloverPC::DslashXpay`, `DiracTwistedCloverPC::M` and `DiracTwistedCloverPC::MdagM` to call the appropriate `Apply*` functions (and combinations of other functions) based on the flavour type. I just don't want to cause conflicts with your local copy.

please go ahead, I can cope with conflicts...

**Comment by urbach** [comment]
If I understand correctly, then what is currently in `dslash_ndeg_twisted_clover_preconditioned` is the symmetric E/O preconditioned form, isn't it?

**Comment by urbach** [comment]
no, okay, no I understand, sorry for the confusion

File: include/dslash_quda.h [hunk 6]
```diff
@@ -532,7 +532,7 @@ namespace quda {
      @param[in] profile The TimeProfile used for profiling the dslash
   */
   void ApplyNdegTwistedCloverPreconditioned(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U,
-      const CloverField &C, double a, double b, double c, bool xpay, const ColorSpinorField &x, int parity, bool dagger,
```
Comment (kostrzewa) [inline include/dslash_quda.h]: I think we need the `xpay` argument at this stage to make the choice in the `apply` function between the different ways of instantiating the kernel, would you agree?

Comment (kostrzewa) [reply include/dslash_quda.h]: I had just forgotten to pull it through all the way to the args.

Comment (urbach) [reply include/dslash_quda.h]: yes, you are right...

**Comment by kostrzewa** [comment]
the issue below has been fixed

## old comment

I think we can't do this (which is the call for the nondeg operator):

```
196   void DiracTwistedCloverPC::Dslash(ColorSpinorField &out, const ColorSpinorField &in, const QudaParity parity) const
197   {
198     checkParitySpinor(in, out);
199     checkSpinorAlias(in, out);
200 
201     if (in.TwistFlavor() != out.TwistFlavor())
202       errorQuda("Twist flavors %d %d don't match", in.TwistFlavor(), out.TwistFlavor());
203     if (in.TwistFlavor() == QUDA_TWIST_NO || in.TwistFlavor() == QUDA_TWIST_INVALID)
204       errorQuda("Twist flavor not set %d", in.TwistFlavor());
205 
206     bool symmetric = (matpcType == QUDA_MATPC_EVEN_EVEN || matpcType == QUDA_MATPC_ODD_ODD) ? true : false;
207     if (dagger && symmetric && !reverse) {
208       printf("TwistCloverInv -> DiracWilson::Dslash\n");
209       bool reset = newTmp(&tmp2, in);
210       TwistCloverInv(*tmp2, in, 1 - parity);
211       DiracWilson::Dslash(out, *tmp2, parity);
212       deleteTmp(&tmp2, reset);
213     } else {
```

as `WilsonApply`, which is instantiated by `ApplyWilson`, which is what is called in `DiracWilson::Dslash`, refers to `in.VolumeCB()`:

```
 35     inline WilsonApply(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a,
 36                        const ColorSpinorField &x, int parity, bool dagger, const int *comm_override, TimeProfile &profile)
 37     {
 38       constexpr int nDim = 4;
 39       WilsonArg<Float, nColor, nDim, recon> arg(out, in, U, a, x, parity, dagger, comm_override);
 40       Wilson<decltype(arg)> wilson(arg, out, in);
 41 
 42       dslash::DslashPolicyTune<decltype(wilson)> policy(
 43         wilson, const_cast<cudaColorSpinorField *>(static_cast<const cudaColorSpinorField *>(&in)), in.VolumeCB(),
 44         in.GhostFaceCB(), profile);
 45     }                                                                                                                                  
 46   };
```

 which is a 5D volume as we've observed previously. I'm not sure if we'll break anything by moving to `volume_4d_cb` here and then I don't know if we need to call the Dslash twice, once for each flavour...

**Comment by kostrzewa** [comment]
Alright, so the `Dslash` test cannot pass for `asymmetric` preconditioning because the host reference code is wrong. The second application of the inverse diagonal term is simply not the correct thing to do.

The asymmetric `M` is implemented by a call to `DiracTwistedCloverPC::Dslash`, followed by a call to `DiracTwistedClover::DslashXpay` with appropriate arguments, such that one ends up with `M_{oo} \psi_o - \kappa^2 D_{oe} M_{ee}^{-1} D_{eo} \psi_o` in the output, as required. As a result, I think I'll be able to wire everything up today with the understanding that the `Dslash` test will continue to fail for the asymmetric preconditioning types, but `MatPC` and company should pass.

**Comment by kostrzewa** [comment]
Things are clearer now after I've wired up all the tests. Unlike for the normal non-deg twisted mass operator, we need `DiracTwistedCloverPC::DslashXpay` as it is required by the non-daggered symmetric PC operator:

```
257   void DiracTwistedCloverPC::M(ColorSpinorField &out, const ColorSpinorField &in) const                                                                  
258   {
259     double kappa2 = -kappa*kappa;
260     bool reset = newTmp(&tmp1, in);
261 
262     bool symmetric =(matpcType == QUDA_MATPC_EVEN_EVEN || matpcType == QUDA_MATPC_ODD_ODD) ? true : false;
263     int odd_bit = (matpcType == QUDA_MATPC_ODD_ODD || matpcType == QUDA_MATPC_ODD_ODD_ASYMMETRIC) ? 1 : 0;
264     QudaParity parity[2] = {static_cast<QudaParity>((1 + odd_bit) % 2), static_cast<QudaParity>((0 + odd_bit) % 2)};
265 
266     if (!symmetric) { // asymmetric preconditioning
267       Dslash(*tmp1, in, parity[0]);
268       DiracTwistedClover::DslashXpay(out, *tmp1, parity[1], in, kappa2);
269     } else if (!dagger) { // symmetric preconditioning
270       Dslash(*tmp1, in, parity[0]);
271       DslashXpay(out, *tmp1, parity[1], in, kappa2);
272     } else { // symmetric preconditioning, dagger
273       TwistCloverInv(out, in, parity[1]);
274       reverse = true;
275       Dslash(*tmp1, out, parity[0]);
276       reverse = false;
277       DiracWilson::DslashXpay(out, *tmp1, parity[1], in, kappa2);
278     }
279 
280     deleteTmp(&tmp1, reset);
281   }
```

**Comment by kostrzewa** [comment]
At this stage I would ask (if possible) for a review of the `tmc_ndeg_matpc` for silly mistakes. Beyond that we'll soon have the tests set up for the non-deg-`DiracTwistedCloverPC::M` (and company) as I've modified the kernel and functor to follow the single-flavour operator more closely (rather than the non-deg twisted mass operator, which is special because it affects how communication is done since the inverse of the diagonal is trivial and can be applied more efficiently than the inverse of the diagonal of the clover operator).

Reviewed this myself and found a bug in the `odd-odd-asym`.

**Comment by kostrzewa** [comment]
If possible a review and merge of #11 would be nice. All the tests for `PC::M` and `PC::MdagM` pass now, such that "only" source preparation and reconstruction remain to be generalised to two flavours.

Beyond this, a review of the flop and byte counters should be done as well as a check of the multi-shift solver to make sure that it supports the two-flavour case. (probably, as it surely must support DWF, but you never know...)

**Comment by kostrzewa** [comment]
I think the source preparation and reconstruction also goes through trivially with the latest commit to #11 thanks to the modifications that I made to `ApplyWilson`.

File: include/blas_helper.cuh [hunk 7]
```diff
@@ -3,6 +3,7 @@
 #include <color_spinor_field.h>
 #include <reduce_helper.h>
 #include <load_store.h>
+#include <convert.h>
```
Comment (urbach) [inline include/blas_helper.cuh]: why is this here?

File: include/gauge_field_order.h [hunk 8]
```diff
@@ -2160,14 +2160,14 @@ namespace quda {
     __device__ __host__ inline void load(complex v[length / 2], int x, int dir, int parity, real = 1.0) const
     {
       // get base pointer
-      auto in = reinterpret_cast<const Float*>(reinterpret_cast<const char*>(gauge) + (parity*volumeCB+x)*size + offset + dir * length);
+      auto in = reinterpret_cast<const Float*>(reinterpret_cast<const char*>(gauge) + (parity*volumeCB+x)*size + offset + dir * length * sizeof(Float));
```
Comment (urbach) [inline include/gauge_field_order.h]: this is coming from `generic kernel`?

File: tests/dslash_test.cpp [hunk 9]
```diff
@@ -611,48 +611,24 @@ void dslashRef()
...
-
-	void *ref1 = spinorRef->V();
-	void *ref2 = (char*)ref1 + tm_offset*cpu_prec;
-    
-	void *flv1 = spinor->V();
-	void *flv2 = (char*)flv1 + tm_offset*cpu_prec;
-    
-	tm_ndeg_dslash(ref1, ref2, hostGauge, flv1, flv2, inv_param.kappa, inv_param.mu, inv_param.epsilon, 
-	               parity, dagger, inv_param.matpc_type, inv_param.cpu_prec, gauge_param);	
+        tm_ndeg_dslash(spinorRef->V(), hostGauge, spinor->V(), inv_param.kappa, inv_param.mu, inv_param.epsilon,
```
Comment (urbach) [inline tests/dslash_test.cpp]: the interface is indeed much much cleaner like this!

File: tests/host_reference/clover_reference.cpp [hunk 10]
```diff
@@ -241,6 +241,68 @@ void twistCloverGamma5(void *out, void *in, void *clover, void *cInv, const int
...
+    applyTwist(out1, in1, tmp1, a, precision);
+    applyTwist(out2, in2, tmp2, -a, precision);
+    // out += (-epsilon tau_1) * in 
+    axpy(b, in2, out1, Vh * spinor_site_size, precision);
+    axpy(b, in1, out2, Vh * spinor_site_size, precision);
+  } else if (twist == QUDA_TWIST_GAMMA5_INVERSE) {
+    void *tmptmp1 = malloc(Vh * spinor_site_size * precision);
+    void *tmptmp2 = malloc(Vh * spinor_site_size * precision);
+
+    a = -2.0 * kappa * mu;
```
Comment (urbach) [inline tests/host_reference/clover_reference.cpp:275]: I don't understand yet, why the sign is different here.

**Comment by kostrzewa** [comment]
Testing `prepare` and `reconstruct` means running (and possibly adjusting) `invert_test` with the correct arguments for `--matpc-type` and `--solution-type`. Complementarily, one could also proceed to extend the tmLQCD QUDA interface to fully support the non-degenerate operators, thus directly testing against tmLQCD.

**Review by urbach — APPROVED** [review]

File: include/blas_helper.cuh (hunk 7 above)
Comment (urbach) [reply include/blas_helper.cuh]: this comes from `generic kernel`

File: tests/host_reference/clover_reference.cpp (hunk 10 above)
Comment (kostrzewa) [reply tests/host_reference/clover_reference.cpp:275]: This is (C - i mu gamma5 tau 3 + eps tau1)/(C^2 + mu^2 - eps^2), whereas for `DIRECT`, we apply (C + i mu gamma5 tau3 - eps tau1).

File: include/gauge_field_order.h (hunk 8 above)
Comment (kostrzewa) [reply include/gauge_field_order.h]: yes

File: include/blas_helper.cuh (hunk 7 above)
Comment (kostrzewa) [reply include/blas_helper.cuh]: yes

**Comment by kostrzewa** [comment]
ugh, latest changes to the generic_kernel branch completely broke our 2-flavour operator... running https://github.com/qcdcode/helper_scripts/blob/master/quda/dslash_test/run_test_NdegTwistedClover.sh I get only failures:

```
Mat_dag-0_matpc-none_rval1
Mat_dag-1_matpc-none_rval1
MatDagMat_dag-0_matpc-none_rval1
MatDagMat_dag-1_matpc-none_rval1
Dslash_dag-0_matpc-odd-odd_rval1
Dslash_dag-0_matpc-even-even_rval1
Dslash_dag-0_matpc-odd-odd-asym_rval1
Dslash_dag-0_matpc-even-even-asym_rval1
Dslash_dag-1_matpc-odd-odd_rval1
Dslash_dag-1_matpc-even-even_rval1
Dslash_dag-1_matpc-odd-odd-asym_rval1
Dslash_dag-1_matpc-even-even-asym_rval1
MatPC_dag-0_matpc-odd-odd_rval1
MatPC_dag-0_matpc-even-even_rval1
MatPC_dag-0_matpc-odd-odd-asym_rval1
MatPC_dag-0_matpc-even-even-asym_rval1
MatPC_dag-1_matpc-odd-odd_rval1
MatPC_dag-1_matpc-even-even_rval1
MatPC_dag-1_matpc-odd-odd-asym_rval1
MatPC_dag-1_matpc-even-even-asym_rval1
MatPCDagMatPC_dag-0_matpc-odd-odd_rval1
MatPCDagMatPC_dag-0_matpc-even-even_rval1
MatPCDagMatPC_dag-0_matpc-odd-odd-asym_rval1
MatPCDagMatPC_dag-0_matpc-even-even-asym_rval1
MatPCDagMatPC_dag-1_matpc-odd-odd_rval1
MatPCDagMatPC_dag-1_matpc-even-even_rval1
MatPCDagMatPC_dag-1_matpc-odd-odd-asym_rval1
MatPCDagMatPC_dag-1_matpc-even-even-asym_rval1
```

**Comment by urbach** [comment]
argh, super... :(

**Comment by kostrzewa** [comment]
The issue is with comms, the operators still pass the tests when run with a single MPI task.

**Comment by kostrzewa** [comment]
https://github.com/lattice/quda/pull/1121
